static/_posts/2024-04-01-vgaterm-intro-001.md

```diff
@@ -0,0 +1,152 @@
+Once the pixel byte has been taken out of the FIFO memory it's on its way to being displayed by the monitor. To be displayed this pixel byte must be converted into three analog voltages between 0 and 0.7 volts corresponding to the Red, Green, and Blue color channels. This is accomplished with a [R2R Resistor Ladder DAC]().
+
+An R2R Resistor Ladder DAC is traditionally connected into an op-amp to maintain the output voltage regardless of load. However in our case the load is just the 75 Ohm impedance of the analog color lines in the monitor, so we really only have the one load. One concern, however, is that the resistor ladder can consume a fair amount of current and the lines representing the color coming from an IC may not be rated to supply the amount of current the DAC would need to maintain the voltage.
+
+In earlier versions of the DAC this is what we did (this is what Ben Eater did in his original "Worlds Worst Video Card" video). We did also try an op-amp but that seemed to create some weird artifacts in the colors in images. Lastly what seemed to work best is just feed the input bit lines into line drivers which _could_ supply the appropriate amount of current and now colors are quite vivid.
+
+### IO Control
+To actually function as a terminal Vgaterm must receive serial signals from a connected host on one of the four IO ports and update the video buffer accordingly. It also must handle keyboard input and forward those typed characters to the connected host.
+
+{% include figure.html image="/assets/images/vgaterm-terminal-sequence.drawio.png" caption="Basic Keyboard input to terminal operation" %}
+
+For keyboard input we went over several options including falling back on [PS/2 protocol over USB](https://en.wikipedia.org/wiki/PS/2_port#Conversion_between_PS/2_and_USB) and trying to write a basic USB host onto the esp32c3. But we settled on just using this [USB host to UART converter](https://www.tindie.com/products/matzelectronics/ch559-usb-host-to-uart-bridge-module/) by [MatzElectronics](https://www.tindie.com/stores/matzelectronics/).
+
+Software, in a loop, waits for keyboard input or input by a connected host. When input is received on either, the CPU is interrupted. If receiving key presses, the character(s) are dequeued and passed along to the connected host. If receiving characters from the host, vgaterm updates the text buffer and cursor which updates the video memory. The updated video memory is then emitted in the next frame, showing the text to the user on screen.
+
+Seth and I decided that this should _never_ interrupt the CPU from emitting pixel bytes as we wanted to prevent any visual artifacts as much as possible. However this means that we could end up more likely to drop characters if the CPU is overwhelmed with input. The CPU UART is configured to issue an `XOFF` character as a part of [Software Flow Control](https://en.wikipedia.org/wiki/Software_flow_control) to tell the host to stop sending input.
```

Comment:
I'd suggest developing this a little bit more on the problem/tradeoff side, and maybe defer XOFF for the follow-up post.

Some important points that feel worth hitting:

* there's hardware buffering & other things that reduce the likelihood of drops on the UART stuff that doesn't exist on the video side (because we didn't build it) 
* frames are ephemeral, and bytes are "forever," but screen shake feels way more like jank (especially given the relative frequencies)
* we uncovered a big difference between average and worst-case terminal processing: and we on purpose decided against over-building for the worst case